pytorch_toolkit/formula_recognition/tools/evaluation_tools.py

```diff
@@ -0,0 +1,392 @@
+    padded_im_1[0:img_data1.shape[0], 0:img_data1.shape[1]] = img_data1
+    padded_im_2[0:img_data2.shape[0], 0:img_data2.shape[1]] = img_data2
+    if (padded_im_1 == padded_im_2).all():
+        return True, True
+
+    # check if difference realy is (e.g. it is not shift on 1-2 px)
+    diff = np.zeros((*padded_im_1.shape, 3), dtype=np.uint8)
+    diff[(padded_im_1 == 1) * (padded_im_2 == 1), :] = (255, 255, 255)
+    diff[(padded_im_1 == 1) * (padded_im_2 == 0), :] = (255, 0, 0)
+    diff[(padded_im_1 == 0) * (padded_im_2 == 1), :] = (0, 0, 255)
+
+    differ = check_differ(diff)
+    if differ:
+        # create color map of differences with spaces
+        cv.imwrite(out_path.replace('.png', '_with_s.png'),
+                   np.transpose(diff, (1, 0, 2)))
+    else:
+        return True, True
+    # remove whitespace colmuns and evaluate images again
+    spaceless_im_1 = np.array(
+        [column for column in padded_im_1 if sum(column == 0) > max_pixel_column_diff])
+    spaceless_im_2 = np.array(
+        [column for column in padded_im_2 if sum(column == 0) > max_pixel_column_diff])
+    if max(spaceless_im_1.shape) == 0 or max(spaceless_im_2.shape) == 0:
+        return False, False
+    if spaceless_im_1.shape == spaceless_im_2.shape and (spaceless_im_1 == spaceless_im_2).all():
+        return False, True
+
+    max_h = max(spaceless_im_1.shape[1], spaceless_im_2.shape[1])
+    max_w = max(spaceless_im_1.shape[0], spaceless_im_2.shape[0])
+    diff = np.zeros((max_w, max_h))
+    mask1 = ((spaceless_im_1 == 1) * 2)
+    mask2 = ((spaceless_im_2 == 1) * 3)
+    diff[0: mask1.shape[0], 0:mask1.shape[1]] += mask1
+    diff[0: mask2.shape[0], 0:mask2.shape[1]] += mask2
+
+    new_diff = np.zeros((*diff.shape, 3), dtype=np.uint8)
+    new_diff[diff == 2] = (255, 0, 0)
+    new_diff[diff == 3] = (0, 0, 255)
+    new_diff[diff == 5] = (255, 255, 255)
+
+    differ = check_differ(new_diff)
+    if differ:
+        cv.imwrite(out_path.replace('.png', '_wout_s.png'),
+                   np.transpose(new_diff, (1, 0, 2)))
+        match_w, match_wout = False, False
+    else:
+        match_w, match_wout = False, True
+    return match_w, match_wout
+
+
+class Im2latexRenderBasedMetric():
+
+    def __init__(self):
+        self.num_threads = cpu_count()
+        self.max_pixel_column_diff = MAX_PX_COL_DIFF
+        check_environment()
+
+    def compare_images(self, images_dir):
+        """
+        Function reads images and compares them, first, as is
+        second, deletes all whitespaces and compares again.
+        This step helps to escape the situation when two pics
+        are not equal because of different length of spaces
+        """
+        total_num = 0
+        total_correct = 0
+        total_correct_eliminate = 0
+        lines = []
+        pool = ThreadPool(self.num_threads)
+        gold_dir = os.path.join(images_dir, 'images_gold')
+        pred_dir = os.path.join(images_dir, 'images_pred')
```

Comment:
IMHO, it is not a good decision generate the strings `gold_dir` and `pred_dir` twice -- once in `render_images`, the second time in `compare_images`. Obviously it is the methods of the same class, but nevertheless it is something like "a hidden dependency".
It is better to add at the end of the method `render_images`
```python
return out_path_gold, out_path_pred
```
And instead of one parameter `images_dir` pass to the method `compare_dir` both of the variables.

Reply:
Added paths as return value to the render function